mockwebserver/src/main/java/okhttp3/mockwebserver/TlsNode.java

```diff
@@ -0,0 +1,142 @@
+ * limitations under the License.
+ */
+package okhttp3.mockwebserver;
+
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.List;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509KeyManager;
+import javax.net.ssl.X509TrustManager;
+import okhttp3.internal.platform.Platform;
+
+import static okhttp3.mockwebserver.internal.tls.TlsUtil.newKeyManager;
+import static okhttp3.mockwebserver.internal.tls.TlsUtil.newTrustManager;
+
+/**
+ * Certificates to identify which peers to trust and also to earn the trust of those peers in kind.
+ * Client and server exchange these certificates during the handshake phase of a TLS connection.
+ *
+ * <h3>Server Authentication</h3>
+ *
+ * <p>This is the most common form of TLS authentication: clients verify that servers are trusted
+ * and that they own the hostnames that they represent. Server authentication is required.
+ *
+ * <p>To perform server authentication:
+ *
+ * <ul>
+ *   <li>The server's TLS node must have a {@linkplain HeldCertificate held certificate} (a
+ *       certificate and its private key). The certificate's subject alternative names must match
+ *       the server's hostname. The server must also have is a (possibly-empty) chain of
+ *       intermediate certificates to establish trust from a root certificate to the server's
+ *       certificate. The root certificate is not included in this chain.
+ *   <li>The client's TLS node must include a set of trusted root certificates. They will be used to
+ *       authenticate the server's certificate chain. Typically this is a set of well-known root
+ *       certificates that is distributed with the HTTP client or its platform. It may be augmented
+ *       by certificates private to an organization or service.
+ * </ul>
+ *
+ * <h3>Client Authentication</h3>
+ *
+ * <p>This is authentication of the client by the server during the TLS handshake. Client
+ * authentication is optional.
+ *
+ * <p>To perform client authentication:
+ *
+ * <ul>
+ *   <li>The client's TLS node must have a {@linkplain HeldCertificate held certificate} (a
+ *       certificate and its private key). The client must also have a (possibly-empty) chain of
+ *       intermediate certificates to establish trust from a root certificate to the client's
+ *       certificate. The root certificate is not included in this chain.
+ *   <li>The server's TLS node must include a set of trusted root certificates. They will be used to
+ *       authenticate the client's certificate chain. Typically this is not the same set of root
+ *       certificates used in server authentication. Instead it will be a small set of roots
+ *       private to an organization or service.
+ * </ul>
+ */
+public final class TlsNode {
```

Comment:
TlsNode may grow on me but initially doesn't tell me what it does.

Reply:
Yeah, 100% agree. I’m looking for better names.